zerver/lib/actions.py

```diff
@@ -3916,26 +3916,29 @@ def do_send_confirmation_email(invitee, referrer, body):
     send_email('zerver/emails/invitation', to_email=invitee.email, from_name=from_name,
                from_address=FromAddress.NOREPLY, context=context)
 
-def user_email_is_unique(email):
-    # type: (Text) -> None
+def email_not_system_bot(email: Text) -> None:
+    if email.lower() in settings.CROSS_REALM_BOT_EMAILS:
+        raise ValidationError('%s is an email address reserved for system bots' % (email,))
+
+def validate_email_for_realm(target_realm: Realm, email: Text) -> None:
     try:
-        get_user_profile_by_email(email)
-        raise ValidationError('%s already has an account' % (email,))
-    except UserProfile.DoesNotExist:
-        pass
+        # Registering with a system bot's email is not allowed...
+        email_not_system_bot(email)
+    except ValidationError:
+        # ... unless this is the first user with that email.
+        if UserProfile.objects.filter(email__iexact=email).exists():
+            raise
```

Comment:
Can you explain why we should allow one to take a system bot's email here?  It seems wrong to me; system bots are never registered directly as a normal user account anyway.

Reply:
Ah, good question. This is what happens in the setup in `zerver.tests.test_messages.TestCrossRealmPMs.test_realm_scenarios`. One of the things that test wanted to test involved a hypothetical system bot in a different realm, so it has to create it, and at that point it's already overridden `CROSS_REALM_BOT_EMAILS` so that that email is for a system bot. This condition allows that setup to go through.

This can probably be fixed on the test side, and that's probably ultimately the right fix. If nothing else I should put a comment here about this, and make a followup task to clean it up.